sw/host/opentitanlib/src/bootstrap/rescue.rs

```diff
@@ -0,0 +1,352 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{anyhow, bail, ensure, Result};
+use mundane::hash::{Digest, Hasher, Sha256};
+use std::time::{Duration, Instant};
+use thiserror::Error;
+use zerocopy::AsBytes;
+
+use crate::app::TransportWrapper;
+use crate::bootstrap::{Bootstrap, BootstrapOptions, UpdateProtocol};
+use crate::io::uart::Uart;
+use crate::transport::Capability;
+
+#[derive(AsBytes, Debug, Default)]
+#[repr(C)]
+struct FrameHeader {
+    hash: [u8; Frame::HASH_LEN],
+    frame_num: u32,
+    flash_offset: u32,
+}
+
+#[derive(AsBytes, Debug)]
+#[repr(C)]
+struct Frame {
+    header: FrameHeader,
+    data: [u8; Frame::DATA_LEN],
+}
+
+impl Default for Frame {
+    fn default() -> Self {
+        Frame {
+            header: Default::default(),
+            data: [0xff; Frame::DATA_LEN],
+        }
+    }
+}
+
+impl Frame {
+    const EOF: u32 = 0x8000_0000;
+    const FLASH_SECTOR_SIZE: usize = 2048;
+    const FLASH_SECTOR_MASK: usize = Self::FLASH_SECTOR_SIZE - 1;
+    const FLASH_BUFFER_SIZE: usize = 128;
+    const FLASH_BUFFER_MASK: usize = Self::FLASH_BUFFER_SIZE - 1;
+    const DATA_LEN: usize = 1024 - std::mem::size_of::<FrameHeader>();
+    const HASH_LEN: usize = 32;
+    const MAGIC_HEADER: [u8; 4] = [0xfd, 0xff, 0xff, 0xff];
+
+    /// Computes the hash in the header.
+    fn header_hash(&self) -> [u8; Frame::HASH_LEN] {
+        let frame = self.as_bytes();
+        let sha = Sha256::hash(&frame[Frame::HASH_LEN..]);
+        sha.bytes()
+    }
+
+    /// Computes the hash over the entire frame.
+    fn frame_hash(&self) -> [u8; Frame::HASH_LEN] {
+        let sha = Sha256::hash(self.as_bytes());
+        let mut digest = sha.bytes();
+        // Touch up zeroes into ones, as that is what the old chips are doing.
+        for b in &mut digest {
+            if *b == 0 {
+                *b = 1;
+            }
+        }
+        digest
+    }
+
+    /// Creates a sequence of frames based on a `payload` binary.
+    fn from_payload(payload: &[u8]) -> Result<Vec<Frame>> {
+        // The given payload will contain up to four sections concatenated together:
+        // RO_A, RW_A optionally follwed by RO_B, RW_B
+        // Each section starts with a magic number on at least a 256 byte boundary.
+
+        // This rescue protocol uses the RW_A section only, which will start at the second
+        // occurrance of the magic value, and end at the third occurrence or at the end of the
+        // file.
+
+        ensure!(
+            &payload[0..4] == &Self::MAGIC_HEADER,
+            RescueError::ImageFormatError
+        );
+
+        // Find second occurrence of magic value.
+        let min_addr = match payload[256..]
+            .chunks(256)
+            .position(|c| &c[0..4] == &Self::MAGIC_HEADER)
```

Comment:
has_prefix, perhaps?

Reply:
As far as I can tell, strings have a starts_with() method, but I have not found an equivalent for array slices, are you proposing that I declare a helper method myself.  I feel that the expression `&c[0..4] == &Self::MAGIC_HEADER` is short enough that a helper method is hard to justify.  (Avoiding to specify the number 4 explicitly should be the only advantage.

Reply:
https://doc.rust-lang.org/std/primitive.slice.html#method.starts_with is what you want. :)

Reply:
Thank you, I do not know how I could miss that.